File: packages/modules/notifications/android/src/main/java/run/stingjs/modules/notifications/NotificationsModule.kt
package run.stingjs.modules.notifications

import android.Manifest
import android.app.Activity
import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import run.stingjs.runtime.StingApplicationLifecycleEvent
import run.stingjs.runtime.StingNativeModule
import run.stingjs.runtime.StingNativeModuleCompletion
import run.stingjs.runtime.StingNativeModuleError
import run.stingjs.runtime.StingNativeModuleEventEmitter
import run.stingjs.runtime.StingNativeModuleResult
import run.stingjs.runtime.StingPermissionCompletion
import run.stingjs.runtime.StingPermissionStatus
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class NotificationsModule(private val context: Context) : StingNativeModule {
    override val name = "Notifications"
    override val version = "0.1.0"
    private val owner = UUID.randomUUID().toString()
    private val alarms = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
    private val prefs = context.getSharedPreferences("sting.notifications", Context.MODE_PRIVATE)

    init { ensureChannel(context) }

    override fun callSync(method: String, arguments: List<Any?>): Any? = throw StingNativeModuleError("E_METHOD_NOT_FOUND", "Notifications does not implement synchronous method $method")
    override fun callAsync(method: String, arguments: List<Any?>, completion: StingNativeModuleCompletion) {
        when (method) {
            "schedule" -> {
                val id = (arguments.getOrNull(0) as? String)?.takeIf { it.isNotBlank() } ?: UUID.randomUUID().toString()
                val title = arguments.getOrNull(1) as? String ?: run { completion(StingNativeModuleResult.Failure(StingNativeModuleError("E_INVALID_ARGUMENT", "Notifications.schedule requires a title."))); return }
                val body = arguments.getOrNull(2) as? String ?: ""
                val at = (arguments.getOrNull(3) as? Number)?.toLong()?.takeIf { it > 0 } ?: System.currentTimeMillis()
                val intent = alarmIntent(context, id, title, body)
                alarms.set(AlarmManager.RTC_WAKEUP, at, intent)
                prefs.edit().putString(id, "$at\n$title\n$body").apply()
                completion(StingNativeModuleResult.Success(id))
            }
            "getScheduled" -> {
                val values = prefs.all.mapNotNull { (id, raw) ->
                    val parts = (raw as? String)?.split('\n', limit = 3) ?: return@mapNotNull null
                    mapOf("id" to id, "at" to (parts.getOrNull(0)?.toDoubleOrNull()), "title" to parts.getOrNull(1).orEmpty(), "body" to parts.getOrNull(2).orEmpty())
                }
                completion(StingNativeModuleResult.Success(values))
            }
            "cancel" -> {
                val id = arguments.firstOrNull() as? String ?: run { completion(StingNativeModuleResult.Failure(StingNativeModuleError("E_INVALID_ARGUMENT", "Notifications.cancel requires an id."))); return }
                alarms.cancel(alarmIntent(context, id, "", "")); prefs.edit().remove(id).apply(); completion(StingNativeModuleResult.Success(null))
            }
            else -> completion(StingNativeModuleResult.Failure(StingNativeModuleError("E_METHOD_NOT_FOUND", "Notifications does not implement asynchronous method $method")))
        }
    }

    override fun setEventEnabled(event: String, enabled: Boolean, emit: StingNativeModuleEventEmitter) {
        if (event != "received" && event != "opened") throw StingNativeModuleError("E_EVENT_NOT_FOUND", "Notifications does not implement event $event")
        if (enabled) NotificationEvents.set(event, owner, emit) else NotificationEvents.remove(event, owner)
    }

    override fun onApplicationLifecycle(event: StingApplicationLifecycleEvent) {
        if (event != StingApplicationLifecycleEvent.RUNTIME_DISPOSING) return
        NotificationEvents.clear(owner)
        NotificationsPermissionActivity.cancel(owner)
    }

    override fun permissionStatus(permission: String): StingPermissionStatus {
        if (permission != "notifications") throw StingNativeModuleError("E_PERMISSION_NOT_FOUND", "Notifications does not implement permission $permission")
        if (Build.VERSION.SDK_INT < 33) return StingPermissionStatus.GRANTED
        return if (context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) StingPermissionStatus.GRANTED else StingPermissionStatus.UNDETERMINED
    }
    override fun requestPermission(permission: String, completion: StingPermissionCompletion) {
        if (permission != "notifications") { completion(Result.failure(StingNativeModuleError("E_PERMISSION_NOT_FOUND", "Notifications does not implement permission $permission"))); return }
        if (Build.VERSION.SDK_INT < 33 || permissionStatus(permission) == StingPermissionStatus.GRANTED) { completion(Result.success(StingPermissionStatus.GRANTED)); return }
        NotificationsPermissionActivity.request(context, owner) { completion(Result.success(permissionStatus(permission))) }
    }

    companion object {
        const val CHANNEL_ID = "sting-default"
        fun ensureChannel(context: Context) { if (Build.VERSION.SDK_INT >= 26) (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).createNotificationChannel(NotificationChannel(CHANNEL_ID, "Sting", NotificationManager.IMPORTANCE_DEFAULT)) }
        fun alarmIntent(context: Context, id: String, title: String, body: String): PendingIntent {
            val intent = Intent(context, NotificationAlarmReceiver::class.java).putExtra("id", id).putExtra("title", title).putExtra("body", body)
            return PendingIntent.getBroadcast(context, id.hashCode(), intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        }
    }
}

private data class OwnedEmitter(val owner: String, val emit: StingNativeModuleEventEmitter)

private object NotificationEvents {
    private val emitters = ConcurrentHashMap<String, OwnedEmitter>()
    fun set(event: String, owner: String, emit: StingNativeModuleEventEmitter) { emitters[event] = OwnedEmitter(owner, emit) }
    fun remove(event: String, owner: String) { emitters[event]?.takeIf { it.owner == owner }?.let { emitters.remove(event, it) } }
    fun clear(owner: String) { emitters.entries.filter { it.value.owner == owner }.forEach { emitters.remove(it.key, it.value) } }
    fun emit(event: String, payload: Map<String, Any?>) { emitters[event]?.emit?.invoke(payload) }
}

class NotificationAlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val id = intent.getStringExtra("id") ?: return
        val title = intent.getStringExtra("title") ?: ""
        val body = intent.getStringExtra("body") ?: ""
        NotificationsModule.ensureChannel(context)
        val opened = PendingIntent.getBroadcast(context, id.hashCode(), Intent(context, NotificationOpenedReceiver::class.java).putExtra("id", id).putExtra("title", title).putExtra("body", body), PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        val notification = if (Build.VERSION.SDK_INT >= 26) android.app.Notification.Builder(context, NotificationsModule.CHANNEL_ID) else @Suppress("DEPRECATION") android.app.Notification.Builder(context)
        notification.setContentTitle(title).setContentText(body).setSmallIcon(android.R.drawable.ic_dialog_info).setContentIntent(opened).setAutoCancel(true)
        (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).notify(id.hashCode(), notification.build())
        context.getSharedPreferences("sting.notifications", Context.MODE_PRIVATE).edit().remove(id).apply()
        NotificationEvents.emit("received", mapOf("id" to id, "title" to title, "body" to body))
    }
}

class NotificationOpenedReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val id = intent.getStringExtra("id") ?: return
        NotificationEvents.emit("opened", mapOf("id" to id, "title" to (intent.getStringExtra("title") ?: ""), "body" to (intent.getStringExtra("body") ?: "")))
    }
}

class NotificationsPermissionActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) { super.onCreate(savedInstanceState); if (savedInstanceState == null && Build.VERSION.SDK_INT >= 33) requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), REQUEST_CODE) else if (Build.VERSION.SDK_INT < 33) { completeAll(); finish() } }
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) { super.onRequestPermissionsResult(requestCode, permissions, grantResults); if (requestCode == REQUEST_CODE) { completeAll(); finish() } }
    companion object {
        private const val REQUEST_CODE = 9045
        private val lock = Any()
        private val callbacks = linkedMapOf<String, () -> Unit>()
        private var active = false
        fun request(context: Context, owner: String, completion: () -> Unit) {
            val shouldLaunch = synchronized(lock) {
                callbacks[owner] = completion
                if (active) false else { active = true; true }
            }
            if (shouldLaunch) context.startActivity(Intent(context, NotificationsPermissionActivity::class.java).apply { if (context !is Activity) addFlags(Intent.FLAG_ACTIVITY_NEW_TASK) })
        }
        fun cancel(owner: String) { synchronized(lock) { callbacks.remove(owner) } }
        private fun completeAll() {
            val pending = synchronized(lock) {
                active = false
                val values = callbacks.values.toList()
                callbacks.clear()
                values
            }
            pending.forEach { it() }
        }
    }
}

File: packages/modules/notifications/ios/NotificationsModule.swift
import Foundation
import StingRuntime
import UserNotifications

public final class NotificationsModule: NSObject, StingNativeModule, UNUserNotificationCenterDelegate {
    public let name = "Notifications"
    public let version = "0.1.0"
    private let center = UNUserNotificationCenter.current()
    private var emitters: [String: StingNativeModuleEventEmitter] = [:]

    public override init() { super.init() }

    public func callSync(method: String, arguments: [Any]) throws -> Any? {
        throw StingNativeModuleError(code: "E_METHOD_NOT_FOUND", message: "Notifications does not implement synchronous method \(method)")
    }

    public func callAsync(method: String, arguments: [Any], completion: @escaping StingNativeModuleCompletion) {
        switch method {
        case "schedule":
            let identifier = ((arguments.first as? String)?.isEmpty == false ? arguments.first as? String : nil) ?? UUID().uuidString
            guard arguments.count > 1, let title = arguments[1] as? String else { completion(.failure(StingNativeModuleError(code: "E_INVALID_ARGUMENT", message: "Notifications.schedule requires a title."))); return }
            let body = arguments.count > 2 ? (arguments[2] as? String ?? "") : ""
            let at = arguments.count > 3 ? (arguments[3] as? NSNumber)?.doubleValue ?? 0 : 0
            let content = UNMutableNotificationContent(); content.title = title; content.body = body; content.sound = .default
            let trigger: UNNotificationTrigger? = at > Date().timeIntervalSince1970 * 1000 + 1000 ? UNTimeIntervalNotificationTrigger(timeInterval: max(1, at / 1000 - Date().timeIntervalSince1970), repeats: false) : nil
            center.add(UNNotificationRequest(identifier: identifier, content: content, trigger: trigger)) { error in
                if let error { completion(.failure(StingNativeModuleError(code: "E_NOTIFICATIONS_SCHEDULE", message: error.localizedDescription))) } else { completion(.success(identifier)) }
            }
        case "getScheduled":
            center.getPendingNotificationRequests { requests in
                completion(.success(requests.map { request in
                    let date = (request.trigger as? UNTimeIntervalNotificationTrigger)?.nextTriggerDate()?.timeIntervalSince1970
                    return ["id": request.identifier, "title": request.content.title, "body": request.content.body, "at": date.map { $0 * 1000 } ?? NSNull()] as [String: Any]
                }))
            }
        case "cancel":
            guard let id = arguments.first as? String else { completion(.failure(StingNativeModuleError(code: "E_INVALID_ARGUMENT", message: "Notifications.cancel requires an id."))); return }
            center.removePendingNotificationRequests(withIdentifiers: [id]); completion(.success(nil))
        default: completion(.failure(StingNativeModuleError(code: "E_METHOD_NOT_FOUND", message: "Notifications does not implement asynchronous method \(method)")))
        }
    }

    public func setEventEnabled(event: String, enabled: Bool, emit: @escaping StingNativeModuleEventEmitter) throws {
        guard event == "received" || event == "opened" else { throw StingNativeModuleError(code: "E_EVENT_NOT_FOUND", message: "Notifications does not implement event \(event)") }
        if enabled { emitters[event] = emit; center.delegate = self } else { emitters.removeValue(forKey: event); if emitters.isEmpty, center.delegate === self { center.delegate = nil } }
    }

    public func applicationLifecycleDidChange(_ event: StingApplicationLifecycleEvent) {
        guard event == .runtimeDisposing else { return }
        emitters.removeAll(keepingCapacity: false)
        if center.delegate === self { center.delegate = nil }
    }

    public func permissionStatus(for permission: String) throws -> StingPermissionStatus {
        guard permission == "notifications" else { throw StingNativeModuleError(code: "E_PERMISSION_NOT_FOUND", message: "Notifications does not implement permission \(permission)") }
        let semaphore = DispatchSemaphore(value: 0)
        var result: StingPermissionStatus = .undetermined
        center.getNotificationSettings { settings in
            switch settings.authorizationStatus {
            case .notDetermined: result = .undetermined
            case .denied: result = .denied
            case .authorized: result = .granted
            case .provisional, .ephemeral: result = .limited
            @unknown default: result = .denied
            }
            semaphore.signal()
        }
        semaphore.wait()
        return result
    }

    public func requestPermission(_ permission: String, completion: @escaping StingPermissionCompletion) {
        guard permission == "notifications" else { completion(.failure(StingNativeModuleError(code: "E_PERMISSION_NOT_FOUND", message: "Notifications does not implement permission \(permission)"))); return }
        center.requestAuthorization(options: [.alert, .badge, .sound]) { [weak self] _, error in
            if let error { completion(.failure(StingNativeModuleError(code: "E_NOTIFICATIONS_PERMISSION", message: error.localizedDescription))); return }
            completion(.success((try? self?.permissionStatus(for: permission)) ?? .denied))
        }
    }

    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        emitters["received"]?(Self.payload(notification.request)); completionHandler([.banner, .sound])
    }
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        emitters["opened"]?(Self.payload(response.notification.request)); completionHandler()
    }
    private static func payload(_ request: UNNotificationRequest) -> [String: Any] { ["id": request.identifier, "title": request.content.title, "body": request.content.body] }
}
